RemoteSpiNNaker/RemoteSpiNNakerJobProcessManager/src/main/java/uk/ac/manchester/cs/spinnaker/jobprocessmanager/JobProcessManager.java

```diff
@@ -100,57 +101,107 @@
         private final Timer sendTimer;
 
         /**
-         * An object to synchronise on when sending data.
+         * The size of the log so far.
          */
-        private final Integer sendSync = new Integer(0);
+        private int logSize;
+
+        /**
+         * The file that the log is being written to locally.
+         */
+        private final File logFile;
+
+        /**
+         * The log file to store logs in as well.
+         */
+        private final FileWriter logFileWriter;
 
         /**
          * Make a log writer that uploads the log every half second.
+         * @throws IOException If there is an error creating a log file.
          */
-        UploadingJobManagerLogWriter() {
+        UploadingJobManagerLogWriter() throws IOException {
             sendTimer = new Timer(UPDATE_INTERVAL, new ActionListener() {
                 @Override
                 public void actionPerformed(final ActionEvent e) {
                     sendLog();
                 }
             });
+            logFile = createTempFile("output_", ".txt", workingDirectory);
+            logFileWriter = new FileWriter(logFile);
+            log("Output will be written to " + logFile.getName());
         }
 
         /**
          * Send the log now if changed.
          */
         private void sendLog() {
-            synchronized (sendSync) {
+            synchronized (this) {
                 String toWrite = null;
-                synchronized (this) {
-                    if (isPopulated()) {
-                        toWrite = takeCache();
-                    }
+                if (isPopulated()) {
+                    toWrite = takeCache();
                 }
                 if (toWrite != null && !toWrite.isEmpty()) {
-                    log("Sending cached data to job manager");
-                    jobManager.appendLog(job.getId(), toWrite);
+                    try {
+                        jobManager.appendLog(job.getId(), toWrite);
+                    } catch (Throwable e) {
+                        log("Error sending log");
+                        log(e);
+                    }
                 }
             }
         }
 
         @Override
-        public void append(final String logMsg) {
-            log("Process Output: " + logMsg);
-            synchronized (this) {
-                sendTimer.restart();
-                appendCache(logMsg);
-                if (cacheSize() >= MAX_LOG_CACHED) {
-                    sendLog();
+        public void append(final String logMsg) throws IOException {
+            logFileWriter.append(logMsg);
+            logFileWriter.flush();
```

Comment:
Is this call really necessary? It's potentially expensive.

Reply:
In general, you only need to `flush` when you're going to hand off to another process without closing (or if you're mad enough to have several streams looking at the same file at once from one process when at least one of those is a writer).

Reply:
This was because the log is no longer written anywhere but here, so flushing here makes it easier to debug when looking at the output on the machine.